packages/nx/src/command-line/migrate.ts

```diff
@@ -487,72 +488,99 @@ function versions(root: string, from: Record<string, string>) {
+      : '';
+
+    if (!packageRegistryCache[packageName]) {
+      const { stdout } = await execAsync(
+        `${pm} config get ${configKeyPrefix}registry`
+      );
+
+      packageRegistryCache[configKeyPrefix] = stdout.trim();
+    }
+    return packageRegistryCache[configKeyPrefix];
   }
 
-  return function nxMigrateFetcher(
+  async function fetchMigrations(
     packageName: string,
     packageVersion: string
   ): Promise<ResolvedMigrationConfiguration> {
-    if (migrationsCache[`${packageName}-${packageVersion}`]) {
-      return migrationsCache[`${packageName}-${packageVersion}`];
+    if (
+      (await getRegistryOfPackage(packageName)) ===
+      'https://npm.pkg.github.com/'
+    ) {
+      /**
+       * The github npm package registry does not yet support custom fields,
+       * meaning that `npm view` will not return the nx migration information,
+       * so we have to fallback to getting the information with a full install.
+       */
+      logger.info(`Fetching ${packageName}@${packageVersion}`);
+      return getPackageMigrationsUsingInstall(packageName, packageVersion);
+    }
+
+    try {
+      const cacheKey = `${packageName}-${packageVersion}`;
+
+      resolvedVersionCache[cacheKey] ??=
+        await resolvePackageVersionUsingRegistry(packageName, packageVersion);
+
+      const resolvedVersion = resolvedVersionCache[cacheKey];
+
+      if (
+        resolvedVersion !== packageVersion &&
+        getMigrationCache(packageName, resolvedVersion)
+      ) {
+        return getMigrationCache(packageName, resolvedVersion);
+      }
+
+      return getPackageMigrationsUsingRegistry(packageName, resolvedVersion);
+    } catch {
+      logger.info(`Fetching ${packageName}@${packageVersion}`);
+      return getPackageMigrationsUsingInstall(packageName, packageVersion);
     }
+  }
 
-    let resolvedVersion: string = packageVersion;
-    let migrations: Promise<ResolvedMigrationConfiguration>;
-    function setCache(packageName: string, packageVersion: string) {
-      migrationsCache[packageName + '-' + packageVersion] = migrations;
+  return async function nxMigrateFetcher(
+    packageName: string,
+    packageVersion: string
+  ): Promise<ResolvedMigrationConfiguration> {
+    if (getMigrationCache(packageName, packageVersion)) {
+      return getMigrationCache(packageName, packageVersion);
     }
-    migrations = fetchMigrations(packageName, packageVersion, setCache).then(
+
+    let migrations = await fetchMigrations(packageName, packageVersion).then(
       (result) => {
         if (result.schematics) {
           result.generators = result.schematics;
           delete result.schematics;
         }
-        resolvedVersion = result.version;
         return result;
       }
     );
 
-    setCache(packageName, packageVersion);
+    setMigrationCache(packageName, packageVersion, migrations);
+    setMigrationCache(packageName, migrations.version, migrations);
```

Comment:
What's the reason behind setting both keys here?

Reply:
This is to replicate the behavior from before. It'd cache the version specified and also the resolved version.